### PR TITLE
Gun websocket

### PR DESCRIPTION
**Pull request by dealloc**
Changes out websocket_client for the :gun library
This allows custom reconnection logic and handling close frames with the error code when Discord closes the gateway.

### CODE REVIEW

File: lib/wumpex/base/websocket.ex [hunk 1]
```diff
@@ -1,202 +1,345 @@
...
+  @callback on_disconnected(state :: term()) :: {term(), reconnect_strategy()}
 
   @doc """
-  Instructs the Websocket to send a given `message`.
+  Called when the socket is reconnected.
 
-  You can pass whether to use binary or text mode using the `options`:
-      # Will use text mode
-      Websocket.send(socket, "Hello world", mode: :text)
+  This funcion takes the current state and returns the new state for the process.
```
Comment (lambstream) [inline lib/wumpex/base/websocket.ex]:
```suggestion
  This function takes the current state and returns the new state for the process.
```

File: lib/wumpex/gateway.ex [hunk 2]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+  @spec send_opcode(gateway :: pid(), opcode :: map()) :: :ok
+  def send_opcode(gateway, opcode) when is_pid(gateway) and is_map(opcode) do
     encoded = :erlang.term_to_binary(opcode)
 
-    Websocket.send(gateway, encoded, mode: :binary)
+    Websocket.send(gateway, {:binary, encoded})
+  end
+
+  @impl Websocket
+  def on_connected(options) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:75]:
```suggestion
  @doc """
  Builds the gateway state
  """
  def on_connected(options) do
```
maybe add some docs here

File: lib/wumpex/gateway.ex [hunk 3]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+      ack: false,
+      sequence: nil,
+      session_id: nil,
+      guild_sup: guild_sup,
+      shard: shard
+    }
+  end
+
+  @impl Websocket
+  def on_disconnected(state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:94]: Add documentation please

File: lib/wumpex/gateway.ex [hunk 4]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+    Logger.info("Disconnected...")
+
+    {{:retry, 1_000}, state}
+  end
+
+  @impl Websocket
+  def on_reconnected(state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:101]: Add documentation please 😄

File: lib/wumpex/gateway.ex [hunk 5]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+    Logger.info("Reconnected!")
+
+    state
+  end
+
+  @impl Websocket
+  def handle_frame({:binary, etf}, state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:108]: Add documentation please 😄

File: lib/wumpex/gateway.ex [hunk 6]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+  # According to the docs, we need to close the connection and resume.
+  @impl GenServer
+  def handle_info({:heartbeat, _interval}, state) do
+    Logger.warn("No heartbeat ack was received between heartbeats!")
+    send_frame(:close)
+
+    {:noreply, %{state | ack: false}}
+  end
+
+  # Handles HELLO
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:159]: Maybe more info on what HELLO is or reference to discord docs?

File: lib/wumpex/gateway.ex [hunk 7]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+  end
+
+  # Handles RESUMED event
+  def dispatch(%{op: 0, s: sequence, t: :RESUMED}, state) do
+    Logger.info("Bot has finished resuming.")
+
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: :GUILD_CREATE, d: event}, %{guild_sup: guild_sup} = state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:204]: Add documentation please 😄

File: lib/wumpex/gateway.ex [hunk 8]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+    Logger.info("Guild became available: #{inspect(event)}")
+
+    {:ok, _guild} = Coordinator.start_guild(guild_sup, event.id)
+
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: :GUILD_DELETE, d: %{id: guild_id}}, state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:214]: Add documentation please 😄

File: lib/wumpex/gateway.ex [hunk 9]
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
...
+    Logger.info("Guild #{guild_id} is no longer available!")
+
+    Coordinator.stop_guild(guild_id)
+
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: event_name, d: %{guild_id: guild_id} = event}, state) do
```
Comment (lambstream) [inline lib/wumpex/gateway.ex:224]: Add documentation please 😄

File: test/support/websocket_helpers.exs [hunk 10]
```diff
@@ -45,3 +45,18 @@ defmodule WebsocketHelpers do
     assert_receive :ready
   end
 end
+
+defmodule WebsocketClient do
+  @moduledoc false
+  use Wumpex.Base.Websocket
+
+  def on_connected(options) do
```
Comment (lambstream) [inline test/support/websocket_helpers.exs:54]: Add documentation please 😄

File: test/support/websocket_helpers.exs [hunk 11]
```diff
@@ -45,3 +45,18 @@ defmodule WebsocketHelpers do
...
+    Keyword.fetch!(options, :worker)
+  end
+
+  def handle_frame(frame, state) do
```
Comment (lambstream) [inline test/support/websocket_helpers.exs:59]: Add documentation please 😄

**Review by lambstream — CHANGES_REQUESTED** [review]

File: lib/wumpex/gateway.ex (hunk 2 above)
Comment (dealloc) [reply lib/wumpex/gateway.ex:75]: Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

File: lib/wumpex/gateway.ex (hunk 3 above)
Comment (dealloc) [reply lib/wumpex/gateway.ex:94]: Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

File: lib/wumpex/gateway.ex (hunk 4 above)
Comment (dealloc) [reply lib/wumpex/gateway.ex:101]: Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

File: lib/wumpex/gateway.ex (hunk 5 above)
Comment (dealloc) [reply lib/wumpex/gateway.ex:108]: Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

**Review by lambstream — APPROVED** [review]

**Pull request merged**